src/components/cursor-controller.js

```diff
@@ -0,0 +1,393 @@
+        direction
+          .set(0, 0, -1)
+          .applyQuaternion(controllerQuaternion)
+          .normalize();
+      }
+
+      this.el.setAttribute("raycaster", { origin: origin, direction: direction });
+
+      let intersection = null;
+
+      //update cursor position
+      if (!this.isGrabbing) {
+        const intersections = this.el.components.raycaster.intersections;
+        if (intersections.length > 0 && intersections[0].distance <= this.data.maxDistance) {
+          intersection = intersections[0];
+          this.data.cursor.object3D.position.copy(intersection.point);
+          this.currentDistance = intersections[0].distance;
+          this.currentDistanceMod = 0;
+        } else {
+          this.currentDistance = this.data.maxDistance;
+        }
+      }
+
+      if (this.isGrabbing || !intersection) {
+        const max = Math.max(this.data.minDistance, this.currentDistance - this.currentDistanceMod);
+        const distance = Math.min(max, this.data.maxDistance);
+        this.currentDistanceMod = this.currentDistance - distance;
+        direction.multiplyScalar(distance);
+        point.addVectors(origin, direction);
+        this.data.cursor.object3D.position.copy(point);
+      }
+
+      //update currentTargetType
+      if (this.isGrabbing && !intersection) {
+        this.currentTargetType = TARGET_TYPE_INTERACTABLE;
+      } else if (intersection) {
+        if (this._isClass("interactable", intersection.object.el)) {
+          this.currentTargetType = TARGET_TYPE_INTERACTABLE;
+        } else if (this._isClass("ui", intersection.object.el)) {
+          this.currentTargetType = TARGET_TYPE_UI;
+        }
+      } else {
+        this.currentTargetType = TARGET_TYPE_NONE;
+      }
+
+      //update cursor material
+      const isTarget = this._isTargetOfType(TARGET_TYPE_INTERACTABLE_OR_UI);
+      if ((this.isGrabbing || isTarget) && !wasIntersecting) {
+        wasIntersecting = true;
+        this.data.cursor.setAttribute("material", { color: this.data.cursorColorHovered });
+      } else if (!this.isGrabbing && !isTarget && wasIntersecting) {
+        wasIntersecting = false;
+        this.data.cursor.setAttribute("material", { color: this.data.cursorColorUnhovered });
+      }
+
+      //update line
+      if (this.hasPointingDevice) {
+        this.el.setAttribute("line", { start: origin.clone(), end: this.data.cursor.object3D.position.clone() });
+      }
+    };
+  })(),
+
+  _isClass: function(className, el) {
+    return (
+      el.className === className ||
+      (el.parentNode && el.parentNode != el.sceneEl && this._isClass(className, el.parentNode))
+    );
+  },
+
+  _isTargetOfType: function(mask) {
+    return (this.currentTargetType & mask) === this.currentTargetType;
+  },
+
+  _startTeleport: function() {
+    const hideCursor = !(this.hasPointingDevice || this.inVR);
+    if (this.controller != null) {
+      this.controller.emit(this.data.teleportEvent, {});
```

Comment:
I'd again consider having this emit a non "action-*" event and instead adding `cursor-teleport_` as a button to the teleport controls, and then emitting `cursor-teleport_down` here.